src/frontend/Semantic_error.ml

```diff
@@ -134,6 +140,22 @@ module TypeError = struct
                   , args )
                 , error ) ]
             , false ) )
+    | AmbiguousFunctionPromotion (s, arg_tys, signatures) ->
+        let pp_sig ppf (rt, args) =
+          Fmt.pf ppf "@[<hov>(@[<hov>%a@]) => %a@]"
+            Fmt.(list ~sep:comma UnsizedType.pp_fun_arg)
+            args UnsizedType.pp_returntype rt in
```

Comment:
This is a bit simplistic. `Signature_mismatch.pp_signature_mismatch` takes extra effort to make higher-order function signatures readable. But of course we don't have user-defined higher-order functions so it probably doesn't matter that much.

Reply:
Yeah, one disadvantage of moving this out of SignatureMismatch was losing that functionality. I suppose we could expose that as a more general function and use it here, but for now I think it’s fine?

That said, if the solution for your failing example is what I suspect, I may have to do it now to get good errors for that. Every time I try to make a simplifying assumption in this PR it’s been wrong, so I should probably stop doing so